File: server/src/security.js
'use strict';
/* Passwords, tokens, validation and abuse limits. No HTTP here. */
const crypto = require('crypto');
const config = require('./config');
const { q } = require('./db');

const scrypt = (pw, salt) => crypto.scryptSync(pw, salt, 64).toString('hex');
const newSalt = () => crypto.randomBytes(16).toString('hex');
const safeEq = (a, b) => a.length === b.length && crypto.timingSafeEqual(Buffer.from(a), Buffer.from(b));
const hashPassword = (pw) => { const salt = newSalt(); return { salt, hash: scrypt(pw, salt) }; };
const checkPassword = (user, pw) => typeof pw === 'string' && safeEq(scrypt(pw, user.salt), user.pass_hash);

const randomPassword = (len = 18) => { const a = 'abcdefghjkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ23456789'; return Array.from(crypto.randomBytes(len), (b) => a[b % a.length]).join(''); };
const randomToken = (bytes = 24) => crypto.randomBytes(bytes).toString('base64url');
const inviteCode = () => crypto.randomBytes(4).toString('hex').toUpperCase();

// ---- signed tokens: {sub|adm, v (token_version), exp} ----
const b64u = (s) => Buffer.from(s).toString('base64url');
const sign = (payload) => `${payload}.${crypto.createHmac('sha256', config.secret).update(payload).digest('base64url')}`;
function verify(token) {
  if (typeof token !== 'string' || !token.includes('.')) return null;
  const [payload, sig] = token.split('.');
  if (!safeEq(sig, crypto.createHmac('sha256', config.secret).update(payload).digest('base64url'))) return null;
  try { const p = JSON.parse(Buffer.from(payload, 'base64url').toString()); return p.exp > Date.now() ? p : null; } catch { return null; }
}
const userToken = (u) => sign(b64u(JSON.stringify({ sub: u.id, v: u.token_version || 0, exp: Date.now() + config.userTokenDays * 864e5 })));
const adminToken = (u) => sign(b64u(JSON.stringify({ adm: u.id, v: u.token_version || 0, exp: Date.now() + config.adminTokenHours * 3600e3 })));
function userFromToken(token) {
  const p = verify(token); if (!p || !p.sub) return null;
  const u = q.users.byId.get(p.sub);
  return u && !u.disabled && (u.token_version || 0) === (p.v || 0) ? u : null;
}
function adminFromToken(token) {
  const p = verify(token); if (!p || !p.adm) return null;
  const u = q.users.byId.get(p.adm);
  return u && u.role === 'admin' && !u.disabled && (u.token_version || 0) === (p.v || 0) ? u : null;
}

// ---- validation ----
const validEmail = (e) => typeof e === 'string' && /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(e) && e.length < 200;
const passwordProblem = (pw) => (typeof pw !== 'string' || pw.length < 8 || pw.length > 200 || !/[0-9]/.test(pw) || !/[a-zA-Z؀-ۿ]/.test(pw)) ? 'weak_password' : null;

// ---- per-IP limiter (15 min window) + exponential tarpit for admin logins ----
const hits = new Map();
function limited(key, max = 30) {
  const now = Date.now(); const rec = hits.get(key) || { n: 0, t: now };
  if (now - rec.t > 15 * 60e3) { rec.n = 0; rec.t = now; }
  rec.n += 1; hits.set(key, rec); return rec.n > max;
}
const fails = new Map();
const tarpit = (ip) => new Promise((r) => setTimeout(r, Math.min(8000, fails.get(ip) ? 500 * 2 ** (fails.get(ip) - 1) : 0)));
const noteFail = (ip) => fails.set(ip, (fails.get(ip) || 0) + 1);
const clearFail = (ip) => fails.delete(ip);
setInterval(() => { const cut = Date.now() - 30 * 60e3; for (const [k, v] of hits) if (v.t < cut) hits.delete(k); }, 10 * 60e3).unref();

module.exports = { hashPassword, checkPassword, randomPassword, randomToken, inviteCode, userToken, adminToken, userFromToken, adminFromToken, validEmail, passwordProblem, limited, tarpit, noteFail, clearFail };

File: backend/internal/app/security.go
package app

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"regexp"
	"strings"
	"sync"
	"time"
	"unicode"

	"golang.org/x/crypto/scrypt"
)

// ---------- passwords (scrypt N=16384, r=8, p=1, 64 bytes; compatible with the previous server) ----------

func hashPassword(pw, salt string) string {
	k, _ := scrypt.Key([]byte(pw), []byte(salt), 16384, 8, 1, 64)
	return hex.EncodeToString(k)
}
func newSalt() string { b := make([]byte, 16); _, _ = rand.Read(b); return hex.EncodeToString(b) }
func checkPassword(hash, salt, pw string) bool {
	return subtle.ConstantTimeCompare([]byte(hashPassword(pw, salt)), []byte(hash)) == 1
}

var emailRe = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

func validEmail(e string) bool { return len(e) < 200 && emailRe.MatchString(e) }

// weakPassword: at least 8 chars with a digit and a letter (Latin or Persian).
func weakPassword(pw string) bool {
	if len(pw) < 8 || len(pw) > 200 {
		return true
	}
	digit, letter := false, false
	for _, r := range pw {
		if unicode.IsDigit(r) {
			digit = true
		} else if unicode.IsLetter(r) {
			letter = true
		}
	}
	return !(digit && letter)
}

func RandomPassword(n int) string {
	const a = "abcdefghjkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ23456789"
	b := make([]byte, n)
	_, _ = rand.Read(b)
	for i := range b {
		b[i] = a[int(b[i])%len(a)]
	}
	return string(b)
}
func randomToken() string { b := make([]byte, 24); _, _ = rand.Read(b); return base64.RawURLEncoding.EncodeToString(b) }
func inviteCode() string   { b := make([]byte, 4); _, _ = rand.Read(b); return strings.ToUpper(hex.EncodeToString(b)) }
func newID() string        { b := make([]byte, 16); _, _ = rand.Read(b); return hex.EncodeToString(b) }

// ---------- signed tokens: base64url(json) . hmac ----------

type tokenClaims struct {
	Sub string `json:"sub,omitempty"` // user session
	Adm string `json:"adm,omitempty"` // admin session
	V   int    `json:"v"`
	Exp int64  `json:"exp"`
}

func signToken(secret []byte, c tokenClaims) string {
	payload, _ := json.Marshal(c)
	p := base64.RawURLEncoding.EncodeToString(payload)
	m := hmac.New(sha256.New, secret)
	m.Write([]byte(p))
	return p + "." + base64.RawURLEncoding.EncodeToString(m.Sum(nil))
}
func parseToken(secret []byte, tok string) (*tokenClaims, bool) {
	i := strings.IndexByte(tok, '.')
	if i <= 0 {
		return nil, false
	}
	p, sig := tok[:i], tok[i+1:]
	m := hmac.New(sha256.New, secret)
	m.Write([]byte(p))
	want := base64.RawURLEncoding.EncodeToString(m.Sum(nil))
	if subtle.ConstantTimeCompare([]byte(sig), []byte(want)) != 1 {
		return nil, false
	}
	raw, err := base64.RawURLEncoding.DecodeString(p)
	if err != nil {
		return nil, false
	}
	var c tokenClaims
	if json.Unmarshal(raw, &c) != nil || c.Exp < time.Now().UnixMilli() {
		return nil, false
	}
	return &c, true
}

// ---------- abuse limits: per-key counter in a 15-minute window + exponential tarpit ----------

type Limiter struct {
	mu    sync.Mutex
	hits  map[string]*hit
	fails map[string]int
}
type hit struct {
	n int
	t time.Time
}

func NewLimiter() *Limiter {
	l := &Limiter{hits: map[string]*hit{}, fails: map[string]int{}}
	go func() {
		for range time.Tick(10 * time.Minute) {
			l.mu.Lock()
			for k, h := range l.hits {
				if time.Since(h.t) > 30*time.Minute {
					delete(l.hits, k)
				}
			}
			l.mu.Unlock()
		}
	}()
	return l
}
func (l *Limiter) Limited(key string, max int) bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	h := l.hits[key]
	if h == nil || time.Since(h.t) > 15*time.Minute {
		h = &hit{t: time.Now()}
		l.hits[key] = h
	}
	h.n++
	return h.n > max
}
func (l *Limiter) Tarpit(key string) time.Duration {
	l.mu.Lock()
	n := l.fails[key]
	l.mu.Unlock()
	if n == 0 {
		return 0
	}
	d := 500 * time.Millisecond << uint(n-1)
	if d > 8*time.Second {
		d = 8 * time.Second
	}
	return d
}
func (l *Limiter) Fail(key string)  { l.mu.Lock(); l.fails[key]++; l.mu.Unlock() }
func (l *Limiter) Clear(key string) { l.mu.Lock(); delete(l.fails, key); l.mu.Unlock() }
